File: web/src/lib/filter.ts
import {
  compactWave,
  fmtBlockers,
  fmtWave,
  paneBackend,
  paneCI,
  paneRuntimeState,
  paneRuntimeTitle,
  prPrimary,
} from "./pane";
import type { PaneView } from "./types";

/* 構造化フィルタ: key:value + 自由語、すべて AND。未知キーは自由語に降格。
 * #filter のテキストが単一の真実 — ドロップダウンは key:value トークンを書き
 * 込むだけで手打ち構文と完全互換。 */

export const FILTER_KEYS = new Set([
  "state",
  "agent",
  "wave",
  "ci",
  "dirty",
  "live",
  "issue",
  "task",
  "pr",
  "run",
  "backend",
]);

export type Term = { kind: "key"; key: string; value: string } | { kind: "word"; word: string };

export function parseQuery(str: string): Term[] {
  const terms: Term[] = [];
  for (const tok of String(str).trim().split(/\s+/)) {
    if (!tok) continue;
    const m = /^([a-z]+):(\S*)$/i.exec(tok);
    if (m && FILTER_KEYS.has(m[1]!.toLowerCase())) {
      terms.push({ kind: "key", key: m[1]!.toLowerCase(), value: m[2]!.toLowerCase() });
    } else {
      terms.push({ kind: "word", word: tok.toLowerCase() });
    }
  }
  return terms;
}

export function matches(p: PaneView, terms: Term[]): boolean {
  const hay =
    p.derived?.filterText ||
    [
      p.issueNum,
      p.taskId,
      p.displayName,
      p.slug,
      p.agent,
      p.branchName,
      p.diffSummary,
      p.dirtyState,
      p.issueState,
      p.paneId,
      paneBackend(p),
      paneRuntimeTitle(p),
      p.agentState,
      fmtWave(p),
      fmtBlockers(p),
    ]
      .join(" ")
      .toLowerCase();
  for (const t of terms) {
    if (t.kind === "word") {
      if (!hay.includes(t.word)) return false;
      continue;
    }
    const pr = prPrimary(p.prs);
    switch (t.key) {
      case "state": {
        const runtimeState = paneRuntimeState(p);
        // runtime 状態に一致するならそれを、さもなくば issue 状態を見る。
        if (
          (runtimeState === t.value ? runtimeState : String(p.issueState ?? "").toLowerCase()) !==
          t.value
        )
          return false;
        break;
      }
      case "run":
        if ((p.derived?.filterValues?.run ?? p.agentState ?? "") !== t.value) return false;
        break;
      case "agent":
        if (
          !String(p.agent ?? "")
            .toLowerCase()
            .includes(t.value)
        )
          return false;
        break;
      case "backend":
        if ((p.derived?.filterValues?.backend ?? paneBackend(p)).toLowerCase() !== t.value)
          return false;
        break;
      case "wave":
        if (
          String(p.wave ?? "") !== t.value &&
          (p.waveLabel ?? "").toLowerCase() !== t.value &&
          (p.derived?.dependencyWave ?? "").toLowerCase() !== t.value &&
          (p.derived?.waveText ?? "").toLowerCase() !== t.value &&
          compactWave(p).toLowerCase() !== t.value &&
          fmtWave(p).toLowerCase() !== t.value
        )
          return false;
        break;
      case "ci":
        if (paneCI(p) !== t.value) return false;
        break;
      case "dirty":
        if (
          (p.derived?.filterValues?.dirty ?? (p.dirtyState === "dirty" ? "yes" : "no")) !== t.value
        )
          return false;
        break;
      case "live":
        if ((p.derived?.filterValues?.live ?? (p.alive ? "yes" : "no")) !== t.value) return false;
        break;
      case "issue":
        if (
          (p.derived?.filterValues?.issue ?? String(p.issueNum)) !== t.value &&
          String(p.issueNum) !== t.value &&
          String(p.taskId ?? "").toLowerCase() !== t.value
        )
          return false;
        break;
      case "task":
        if ((p.taskId ?? "").toLowerCase() !== t.value) return false;
        break;
      case "pr":
        if ((p.derived?.filterValues?.pr ?? pr?.state ?? "none").toLowerCase() !== t.value)
          return false;
        break;
    }
  }
  return true;
}

/* ---- フィルタ文字列のトークン操作(ドロップダウン+チップ UI 用) ---- */

export function filterTokens(filter: string): string[] {
  return filter.trim().split(/\s+/).filter(Boolean);
}

/* key: プレフィックス一致(大文字小文字無視)。replaceToken / stripKey /
 * tokenForKey は必ずこの 1 つの判定を共有する(判定が割れると trigger の
 * 点灯・チェック表示・チップ行が食い違う)。 */
function hasKey(tok: string, key: string): boolean {
  return tok.toLowerCase().startsWith(`${key}:`);
}

/* 同キーの既存トークンを置き換えて追加(state:open → state:closed は上書き) */
export function replaceToken(tokens: string[], key: string, value: string): string[] {
  const toks = stripKey(tokens, key);
  toks.push(`${key}:${value}`);
  return toks;
}

/* 指定キーのトークンを全て除去(手打ちの重複・大文字違いも残さない)。
 * ドロップダウンのトグルオフ用 — exact-match の removeToken だと
 * 「state:open STATE:CLOSED」のような手打ち重複の取り残しが出る。 */
export function stripKey(tokens: string[], key: string): string[] {
  return tokens.filter((t) => !hasKey(t, key));
}

export function removeToken(tokens: string[], tok: string): string[] {
  return tokens.filter((t) => t !== tok);
}

/* 指定キーの最初の key:value トークンの値(小文字)。ドロップダウンの
 * アクティブ表示・トグルオフ判定用。該当キーが無ければ null。 */
export function tokenForKey(tokens: string[], key: string): string | null {
  for (const t of tokens) {
    if (hasKey(t, key)) return t.slice(key.length + 1).toLowerCase();
  }
  return null;
}

File: internal/ui/tui/filter.go
package tui

import (
	"strconv"
	"strings"

	tea "github.com/charmbracelet/bubbletea"
)

type paneFilter struct {
	terms    []string
	states   []string
	agents   []string
	waves    []string
	runs     []string
	cis      []string
	dirty    []string
	live     []string
	issues   []string
	prs      []string
	tasks    []string
	backends []string
}

func (m model) updateFilterInput(msg tea.KeyMsg) (tea.Model, tea.Cmd) {
	switch msg.String() {
	case "ctrl+c":
		return m.quit()
	case "enter", "esc":
		m.filterEditing = false
		return m, nil
	case "backspace", "ctrl+h":
		m.filterQuery = trimLastRune(m.filterQuery)
	case "ctrl+u":
		m.filterQuery = ""
	default:
		if len(msg.Runes) == 0 {
			return m, nil
		}
		m.filterQuery += string(msg.Runes)
	}
	m.refreshRows()
	return m, nil
}

func filterPaneViews(panes []paneView, query string) []paneView {
	filter := parsePaneFilter(query)
	if filter.empty() {
		return panes
	}
	out := make([]paneView, 0, len(panes))
	for _, pane := range panes {
		if pane.matchesFilter(filter) {
			out = append(out, pane)
		}
	}
	return out
}

func parsePaneFilter(query string) paneFilter {
	var filter paneFilter
	for token := range strings.FieldsSeq(strings.ToLower(strings.TrimSpace(query))) {
		key, value, ok := splitFilterToken(token)
		if !ok {
			filter.terms = append(filter.terms, token)
			continue
		}
		switch key {
		case "state", "status", "s":
			filter.states = append(filter.states, value)
		case "agent", "a":
			filter.agents = append(filter.agents, value)
		case "wave", "w":
			filter.waves = append(filter.waves, value)
		case "run":
			filter.runs = append(filter.runs, value)
		case "ci":
			filter.cis = append(filter.cis, value)
		case "dirty":
			filter.dirty = append(filter.dirty, value)
		case "live":
			filter.live = append(filter.live, value)
		case "issue":
			filter.issues = append(filter.issues, value)
		case "pr":
			filter.prs = append(filter.prs, value)
		case "task", "t":
			filter.tasks = append(filter.tasks, value)
		case "backend", "b":
			filter.backends = append(filter.backends, value)
		default:
			filter.terms = append(filter.terms, token)
		}
	}
	return filter
}

func splitFilterToken(token string) (string, string, bool) {
	idx := strings.IndexAny(token, ":=")
	if idx <= 0 || idx == len(token)-1 {
		return "", "", false
	}
	key := strings.TrimSpace(token[:idx])
	value := strings.TrimSpace(token[idx+1:])
	if key == "" || value == "" {
		return "", "", false
	}
	return key, value, true
}

func (f paneFilter) empty() bool {
	return len(f.terms) == 0 && len(f.states) == 0 && len(f.agents) == 0 && len(f.waves) == 0 &&
		len(f.runs) == 0 && len(f.cis) == 0 && len(f.dirty) == 0 && len(f.live) == 0 &&
		len(f.issues) == 0 && len(f.prs) == 0 && len(f.tasks) == 0 && len(f.backends) == 0
}

func (p paneView) matchesFilter(filter paneFilter) bool {
	for _, state := range filter.states {
		if !containsFold(state, p.IssueState, p.TmuxState, p.PRSummary) {
			return false
		}
	}
	for _, agent := range filter.agents {
		if !containsFold(agent, p.Agent) {
			return false
		}
	}
	for _, wave := range filter.waves {
		if !containsFold(wave, p.WaveLabel, p.WaveBadge, p.waveCell(), p.dependencyWaveText()) {
			return false
		}
	}
	for _, run := range filter.runs {
		if !equalFold(run, p.AgentState) {
			return false
		}
	}
	for _, ci := range filter.cis {
		if !equalFold(ci, p.paneCI()) {
			return false
		}
	}
	for _, dirty := range filter.dirty {
		if !equalFold(dirty, yesNo(p.DirtyState == "dirty")) {
			return false
		}
	}
	for _, live := range filter.live {
		if !equalFold(live, yesNo(p.TmuxState == "live")) {
			return false
		}
	}
	for _, issue := range filter.issues {
		value := strings.TrimSpace(p.Derived.FilterValues["issue"])
		if value == "" {
			value = strconv.Itoa(p.IssueNum)
			if p.IssueNum <= 0 && strings.TrimSpace(p.TaskID) != "" {
				value = strings.ToLower(strings.TrimSpace(p.TaskID))
			}
		}
		if !equalFold(issue, value) {
			return false
		}
	}
	for _, pr := range filter.prs {
		if !equalFold(pr, p.primaryPRState()) {
			return false
		}
	}
	for _, task := range filter.tasks {
		if !containsFold(task, p.TaskID) {
			return false
		}
	}
	for _, runtimeBackend := range filter.backends {
		if !equalFold(runtimeBackend, p.backendLabel()) {
			return false
		}
	}
	searchText := strings.ToLower(strings.Join([]string{
		p.Parent,
		p.TaskID,
		p.Kind,
		p.itemLabel(),
		"#" + strconv.Itoa(p.IssueNum),
		strconv.Itoa(p.IssueNum),
		p.TaskID,
		p.Name,
		p.PaneID,
		p.backendLabel(),
		p.TmuxState,
		p.TmuxTitle,
		p.AgentState,
		p.IssueState,
		p.PRSummary,
		p.CIStatus,
		p.BranchName,
		p.WorktreePath,
		p.DiffSummary,
		p.DirtyState,
		p.WorktreeErr,
		p.Agent,
		p.WaveLabel,
		p.WaveBadge,
		p.Blockers,
		p.dependencyWaveText(),
		p.CreatedAt,
		p.Prompt,
	}, "\n"))
	if strings.TrimSpace(p.Derived.FilterText) != "" {
		searchText += "\n" + p.Derived.FilterText
	}
	for _, term := range filter.terms {
		if !strings.Contains(searchText, term) {
			return false
		}
	}
	return true
}

func equalFold(a, b string) bool {
	return strings.EqualFold(strings.TrimSpace(a), strings.TrimSpace(b))
}

func yesNo(v bool) string {
	if v {
		return "yes"
	}
	return "no"
}

func containsFold(needle string, values ...string) bool {
	needle = strings.ToLower(strings.TrimSpace(needle))
	if needle == "" {
		return true
	}
	for _, value := range values {
		if strings.Contains(strings.ToLower(value), needle) {
			return true
		}
	}
	return false
}
